### DISCUSSION TITLE
how to set a loglevel and how to alter parameters?

### DISCUSSION THREAD

**Discussion by olisnr** [Xyce/Xyce#109] [General]
hi
i know ngspice, but want to use Xyce. there are differences i dont know to address. as example it seems, that Xyce dont know `alterparam`. what is the way to change parameters in a simulation.
and the log level is a bit low. my simulation fails with:

```
*****
***** Welcome to the Xyce(TM) Parallel Electronic Simulator
*****
***** This is version Xyce DEVELOPMENT-202409220015-(Public_Release-7.8.0-128-g5b74b0d5)-opensource
***** Date: Tue Oct 22 22:52:10 CEST 2024


***** Executing netlist /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_mostt.spice

***** Reading and parsing netlist...
Netlist error: No analysis specified.
Simulation aborted due to error.  There are 0 MSG_FATAL errors and 1 MSG_ERROR
 errors
*** Xyce Abort ***
```

how can i find out, what was wrong? the netlist was:

```
** sch_path: /ALL/Xschem/OTA33_xyce_test/OTA33_test_xyce.sch
**.subckt OTA33_test_xyce
VDD VDD GND 1.65
VPLUS1 net2 MINUS 50m
VPLUS2 MINUS net1 50m
VPLUS3 net4 MINUS 50m
VPLUS4 MINUS net3 50m
VPLUS PLUS GND 0
VMINUS MINUS GND 0 pulse -100m 100m 1u 10n 10n 33u 66u ac 1 0
VSS GND VSS 1.65
VSSac VSS VSSb 0 ac 1 0
VDDac1 VDDa VDD 0 ac 1 0
x7 VDD MINUS follow+ VSS follow+ net5 follow+ follow+ follow+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
x1 VDD PLUS MINUS VSS in+ net6 in+ in+ in+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
x2 VDD MINUS PLUS VSS ip+ net7 ip+ ip+ ip+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
x3 VDD net1 net2 VSS cmn+ net8 cmn+ cmn+ cmn+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
x4 VDD net4 net3 VSS cmp+ net9 cmp+ cmp+ cmp+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
x6 VDDa GND GND VSS pdd+ net10 pdd+ pdd+ pdd+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
x8 VDD GND GND VSSb pss+ net11 pss+ pss+ pss+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
I0 net5 GND {iset}
I1 net6 GND {iset}
I4 net9 GND {iset}
I6 net10 GND {iset}
I7 net11 GND {iset}
I3 net8 GND {iset}
I2 net7 GND {iset}
x9 VDD GND net12 VSS invert+ net13 invert+ invert+ invert+ OTA33 nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
I5 net13 GND {iset}
R9 net12 MINUS 100k
R10 net12 net14 100k
E1 net14 GND invert+ GND 1
**** begin user architecture code





    * schematic: OTA33_test_xyce
    * dir: /ALL/Xschem/OTA33_xyce_test
    * test: /ALL/Xschem/OTA33_xyce_test/OTA33_BiAS.sym

    * mos_corner:

    .include ~/share/pdk/ihp-sg13g2/libs.tech/ngspice/models/diodes.lib

    .lib ~/share/pdk/ihp-sg13g2/libs.tech/ngspice/models/cornerRES.lib     res_typ
    .lib ~/share/pdk/ihp-sg13g2/libs.tech/ngspice/models/cornerMOShv.lib   mos_tt
    .lib ~/share/pdk/ihp-sg13g2/libs.tech/ngspice/models/cornerMOSlv.lib   mos_tt
    .lib ~/share/pdk/ihp-sg13g2/libs.tech/ngspice/models/cornerCAP.lib     cap_typ
    .lib ~/share/pdk/ihp-sg13g2/libs.tech/ngspice/models/cornerHBT.lib     hbt_typ

    .option rshunt = 1.0e9
    .option rseries = 1.0e-6

    .param nw=1e-6
    .param nl=1e-6
    .param pw=2e-6
    .param pl=1e-6

    .param iset=1u

        * foreach begin

                *alterparam iset = 1u

               .dc vminus -3 3 0.1
               .print format=raw file=write /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_dc_mostt.raw
               .dc vminus -10m 10m 100u
               .print format=raw file=write /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_zoom_dc_mostt.raw
               .tran 1n 77u
               .print format=raw file=write /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_mostt.raw

                *alterparam iset = 10u

               .dc vminus -3 3 0.1
               .print format=raw file=write /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_dc_mostt.raw
               .dc vminus -10m 10m 100u
               .print format=raw file=write /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_zoom_dc_mostt.raw
               .tran 1n 77u
               .print format=raw file=write /ALL/Xschem/OTA33_xyce_test/simulation/OTA33_test_xyce_mostt.raw

        * end

        quit
    .endc





**** end user architecture code
**.ends

* expanding   symbol:  /ALL/Xschem/OTA33_xyce_test/OTA33.sym # of pins=9
** sym_path: /ALL/Xschem/OTA33_xyce_test/OTA33.sym
** sch_path: /ALL/Xschem/OTA33_xyce_test/OTA33.sch
.subckt OTA33 VDD ip in VSS op sink C1 C3 C2  nw=1u nl=1u pw=2u pl=1u
*.ipin in
*.iopin VSS
*.iopin VDD
*.iopin sink
*.ipin ip
*.opin op
*.iopin C1
*.iopin C2
*.iopin C3
XM1 net10 net18 net2 VDD sg13_hv_pmos w=2.4e-05 l=2e-06 ng=5 m=2
XM2 net9 net17 net2 VDD sg13_hv_pmos w=2.4e-05 l=2e-06 ng=5 m=2
Vmeas3 VDD net4 0
XM3 net10 net10 net1 VSS sg13_hv_nmos w=1e-06 l=1e-06 ng=1
XM4 net7 net10 net3 VSS sg13_hv_nmos w=2e-06 l=1e-06 ng=2
Vmeas8 net3 VSS 0
Vmeas9 net1 VSS 0
XM5 net8 V++ net4 VDD sg13_hv_pmos w=8e-06 l=1e-06 ng=2
**** begin user architecture code


**** end user architecture code
XM6 net2 V+ net8 VDD sg13_hv_pmos w=8e-06 l=1e-06 ng=2
XM7 net16 net9 net6 VSS sg13_hv_nmos w=2e-06 l=1e-06 ng=2
XM8 net9 net9 net5 VSS sg13_hv_nmos w=1e-06 l=1e-06 ng=1
Vmeas10 net5 VSS 0
Vmeas11 net6 VSS 0
XM9 net7 net17 net2 VDD sg13_hv_pmos w=2.4e-05 l=2e-06 ng=5 m=2
XM10 net16 net18 net2 VDD sg13_hv_pmos w=2.4e-05 l=2e-06 ng=5 m=2
Vmeas2 VDD net11 0
XM11 net14 net15 net11 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=1
XM12 op V+ net14 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=1
XM13 op V- net7 VSS sg13_hv_nmos w=1.5975e-05 l=1e-06 ng=9
Vmeas5 VDD net12 0
XM14 net13 net15 net12 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=1
XM15 net15 V+ net13 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=1
XM16 net15 V- net16 VSS sg13_hv_nmos w=1.5975e-05 l=1e-06 ng=9
x12 net19 sink V++ V+ V- V-- net20 OTA33_BiAS nw=1e-06 nl=1e-06 pw=2e-06 pl=1e-06
Vmeas1 net20 VSS 0
Vmeas4 VDD net19 0
R14 ip net18 100
R11 net17 in 100
XC2 C3 VSS cap_cmim w=9.0e-6 l=30.0e-6
XC3 C2 VSS cap_cmim w=9.0e-6 l=30.0e-6
XC4 C1 VSS cap_cmim w=9.0e-6 l=30.0e-6
.ends


* expanding   symbol:  /ALL/Xschem/OTA33_xyce_test/OTA33_BiAS.sym # of pins=7
** sym_path: /ALL/Xschem/OTA33_xyce_test/OTA33_BiAS.sym
** sch_path: /ALL/Xschem/OTA33_xyce_test/OTA33_BiAS.sch
.subckt OTA33_BiAS VDD sink V++ V+ V- V-- VSS  nw=1u nl=1u pw=2u pl=1u
*.iopin VSS
*.iopin VDD
*.iopin sink
*.iopin V++
*.iopin V+
*.iopin V-
*.iopin V--
XM1 net4 V+ net2 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=2
Vmeas4 VDD net7 0
Vmeas6 net1 VSS 0
XM2 net3 V-- net1 VSS sg13_hv_nmos w=1e-06 l=1e-06 ng=1
XM3 V-- V- net3 VSS sg13_hv_nmos w=1e-06 l=1e-06 ng=1
XM4 net2 V++ net7 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=2
XM5 V- net4 V-- VSS sg13_hv_nmos w=1e-06 l=1e-06 ng=1
XM6 net4 net4 V- VSS sg13_hv_nmos w=1e-06 l=2e-06 ng=1
XM7 V++ V+ net5 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=2
XM8 net5 V++ net6 VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=2
XM9 sink sink V+ VDD sg13_hv_pmos w=2e-06 l=2e-06 ng=1
XM10 V+ sink V++ VDD sg13_hv_pmos w=2e-06 l=1e-06 ng=2
Vmeas9 VDD net6 0
.ends

.GLOBAL GND
.end
```

**Comment by tvrusso** [comment]
I can't give you a complete answer here, but Xyce and ngspice are not perfectly compatible with each other and their netlist language is somewhat different.  Xyce, like ngspice, has some similarity to SPICE3F5 (ngspice because it has its origins in SPICE3F5, and Xyce, while written from scratch, had as an early design requirement that it must be mostly compatible with SPICE3F5), but both have diverged significantly from that original basic netlist spec.

While basic device lines work in mostly the same way (modulo some tricky issues about level numbers and model version selection), there are other important differences that you must be aware of. 

One important difference is that Xyce, like the original SPICE before it, treats netlist lines that start with a space as comments --- so your attempt at indentation means that all your analysis and .print lines are being ignored.  That's why you're getting the "no analysis specified" error even though you have multiple ".dc" lines.  Freeform indentation of netlists is not supported in Xyce.  Remove the leading spaces and that will stop, but your use of multiple analysis statements (multiple .dc and multiple .tran in the same netlist) will cause other problems.  You can only have one per run, so you can't do both a DC sweep and a transient run in the same pass, you'll have to split them up into multiple netlists.

Xyce does not support "alterparam", as you have noticed.  It does, however, support "stepping" over parameters using the ".STEP" directive.   In your case, since you appear to be trying to run two simulations at two specific values of "iset", you would probably want to use the LIST form of the .STEP directive (see section 2.1.36, starting on page 153 of the Xyce 7.8 reference guide, available at the Xyce web site's [documentation and tutorials](https://xyce.sandia.gov/documentation-tutorials/) section).  Using .step is sort of like having a for loop around the analysis statement, and that analysis will be repeated at each step after modifying the parameters listed.   You can have more than one .step, they nest.

Your print lines are not compatible with Xyce.  You will have to modify them to be Xyce-compatible:
  - all `.PRINT` lines must specify which analysis they apply to (DC in your case)
  - your use of "file=write" followed by a filename isn't what Xyce expects.  Xyce expects "file=" to be followed by the file name directly 
  - `.PRINT` lines must also have a list of things to output, even when it's a raw format.  Your print lines have no outputs listed on them.

See section 2.1.31 of the reference guide for `.PRINT` syntax.

However, if you want a SPICE-style rawfile that contains *all* solution values (rather than a hand-picked subset), you must ask for a raw file on the command line using the "-r" command line option followed by the filename.   See page 769 of the reference guide for command line options.  In this case, you can simply have no `.PRINT` statement in the netlist.

Lastly, I do not know if the PDK you're using is compatible with Xyce.  While the team has worked hard to make Xyce able to support more and more of the PDKs available, many remain incompatible as distributed and must be converted for use in Xyce.  Unfortunately, the "XDM" tool that the team provides is not specificially designed to translate netlists and PDKs from ngspice to Xyce format, as it is focused on conversions from other commercial tools like HSPICE.  If your PDK was designed for use in HSPICE and just happens to work unmodified in ngspice, you might be able to use XDM to convert it.


Verbosity (logging) is normally not compiled in to the code, as the way it was written it was made more for the use of developers than for users.  As such, if you really want verbose output you have to add additional flags to your `configure` invocation to compile in that feature, and the output you get will then be _voluminous_  and may or may not even be useful to you.  

Since your issues here are specifically in netlist syntax, you're better off not going down the verbose build path.

**Reply by tvrusso** [reply]
Couple of other things:
    - .endc is an ngspice thing, and Xyce won't recognize it.  It's not complaining yet because that line happens to start with spaces and so is being ignored, but it's gotta go.
    - "quit" is not a valid Xyce line, so that, too, has to go and is only not causing problems yet because it's indented.

Remove those two lines and all of the attempts at indentation elsewhere in the file, and you will get a little farther.  But all of the things I listed in the previous comment have to be fixed as well.

**Comment by olisnr** [comment]
Thanks a lot! You are very help full. 

i think Your answer should be added to the Xyce_Reference_Guide, there is a section about differences to other spice variants, but this things are missing.
